Доработал Filmorate добавив ветку add-database с функциональностью сохранения состояния данных после перезапуска.

ViktorSalk opened the pull request:
1) Добавил в проект зависимости com.h2database.h2
2) Сконфигурировал базу данных для рабочего режима с помощью файла настроек application.properties. 
3) Сформировал базу данных и собрал SQL запросы в schema.sql и данные в data.sql 
4) Создал DAO для получения данных по фильмам и юзерам из базы.Реализовав имплементацию UserDbStorage и FilmDbStorage с помощью аннотаций @Qualifier. 
5) Для самопроверки создал DataPersistenceTest с уникальными данными, сохраняемыми после перезапуска. 
6) Реализовал CRUD-операции с использованием JdbcTemplate. 
7) Доработал бизнес логику.
8) Обновил старые тесты под доработанную логику.
9) Добавил тесты на интеграционное тестирование через @AutoConfigureTestDatabase.

Viktor3642 committed:
Доработал Filmorate добавив ветку add-database с функциональностью сохранения состояния данных после перезапуска.
1) Добавил в проект зависимости com.h2database.h2
2) Сконфигурировал базу данных для рабочего режима с помощью файла настроек application.properties.
3) Сформировал базу данных и собрал SQL запросы в schema.sql и данные в data.sql
4) Создал DAO для получения данных по фильмам и юзерам из базы.Реализовав имплементацию UserDbStorage и FilmDbStorage с помощью аннотаций @Qualifier.
5) Для самопроверки создал DataPersistenceTest с уникальными данными, сохраняемыми после перезапуска.
6) Реализовал CRUD-операции с использованием JdbcTemplate.
7) Доработал бизнес логику.
8) Обновил старые тесты под доработанную логику.
9) Добавил тесты на интеграционное тестирование через @AutoConfigureTestDatabase.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -25,42 +27,56 @@
     }
-}
+
+    private void validateMpaAndGenres(Film film) {
```
По своей сути контроллер только обрабатывает (маршрутизирует сервисам) запросы, а тут же в контроллере выполняется некоторая бизнес-логика, ее лучше перенести в сервис, иначе логика размазывается по нескольким классам :)
Допустим сервис был вызван из другого сервиса, тогда у нас логика будет другой - без валидации, вряд ли это правильное поведение, лучше когда все работает единообразно

ViktorSalk replied:
Спасибо) Перенес логику проверки validateMpaAndGenres из FilmController в FilmService приведя к большему единообразию.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDbStorage.java:
```diff
@@ -0,0 +1,182 @@
+        if (film.getGenres() != null && !film.getGenres().isEmpty()) {
+            String sql = "INSERT INTO film_genres (film_id, genre_id) VALUES (?, ?)";
+            for (GenreDto genre : film.getGenres()) {
+                jdbcTemplate.update(sql, film.getId(), genre.getId());
```
Запрос в БД - дорогая операция, их лучше минимизировать, а тут же в цикле идут запросы для каждого элемента. Это не особо эффективно особенно если у нас бтудет много записей в БД.
Можно слепить все инсерты в одну строку и выполнить ее за раз, ак будет гораздо эффективней с точки зрения производительности

ViktorSalk replied:
В FilmDbStorage оптимизировал updateFilmGenres для использования пакетных операций вместо отдельных запросов в цикле.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDbStorage.java:
```diff
@@ -0,0 +1,182 @@
+
+    private void loadGenresForFilms(List<Film> films) {
+        for (Film film : films) {
+            loadGenresForFilm(film);
```
Тут бы тоже от запосов в циуле избавится 
Запрос будет что то вроде ... WHERE fg.film_id in (ids)

ViktorSalk replied:
В FilmDbStorage loadGenresForFilms теперь используется один запрос с предложением IN вместо нескольких запросов.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,108 +1,62 @@
-                .filter(Objects::nonNull)
-                .collect(Collectors.toList());
+    public void addFriend(Long userId, Long friendId) {
+        userStorage.get(userId);
```
userStorage.get(userId);

А что тут происходит не совсем понятно, предположу что проверка на существование записи, может тогда завести метод exist который возвращает boolean, код должен стать наглядней и проще :)

ViktorSalk replied:
Спасибо ), так и есть: здесь была проверка на существование записи; в UserService внедрил использование метода exists() для более четкой и эффективной проверки существования пользователя.

EugeneLenkevich replied:
Тут такая история. На уровне сервиса мы оперируем терминами бизнес-логики, переменные - это наши существительные, а методы - глаголы. В таком виде код читается как простой текст-описание какой то бизнесовой задачи. Такой код потом гораздо проще поддерживать. Потому тут лучше задуматься о семантике :)

EugeneLenkevich reviewed (changes requested):
Привет, Виктор! Спасибо за PR.

Отличная работа! :)
В первую очередь попробуй избавиться от логики в контроллерах (это архитектурно не очень правильно) и от запросов в циклах (потенциальная проблема с производительностью, N+1 запрос) по возможности, отметил в комментариях ниже, посмотри, пожалуйста

Viktor3642 committed:
Доработал ветку add-database. Теперь код стал более производительным и более удобен в обслуживании.
1) Перенес логику проверки validateMpaAndGenres из FilmController в FilmService приведя к большему единообразию.
2) В FilmDbStorage оптимизировал updateFilmGenres для использования пакетных операций вместо отдельных запросов в цикле.
3) В FilmDbStorage loadGenresForFilms теперь используется один запрос с предложением IN вместо нескольких запросов.
4) В UserService внедрил использование метода exists() для более четкой и эффективной проверки существования пользователя.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/dao/UserDbStorage.java:
```diff
@@ -93,12 +92,18 @@ public List<User> getFriends(Long userId) {
+
+    @Override
+    public boolean exists(Long userId) {
+        String sql = "SELECT COUNT(*) FROM users WHERE user_id = ?";
```
А есть еще отдельный exists-запрос

https://stackoverflow.com/questions/7471625/fastest-check-if-row-exists-in-postgresql

EugeneLenkevich reviewed (approved):
Теперь все отлично, спасибо 

PS дальше у нас уже будет Spring Data, запросы станет писать и раскладывать результаты по объектам гораздо проще :)
